From: Rene Kita <mail@rkta.de>
Date: Sun Apr 28 08:11:05 CEST 2024
Subject: Fix segfault with dump_source on compressed data
Patch: patches/Fix-segfault-with-dump_source-on-compressed-data.patch

When one of -dump_source, -dump_both or -dump_extra is set and the
server returns compressed data the load function loadproc does not get
set because the content type for for compressed data is not handled in
the if-else block. This leads to reading of uninitialized memory when
the function is called through the pointer.

As of today I don't understand what should be done to dump in this case,
but a segfault is certainly the wrong result.

Set loadproc to NULL, which will lead to an error message, that the site
could not be loaded.
---
 file.c |    9 +++++++++
 1 file changed, 9 insertions(+)

--- a/file.c
+++ b/file.c
@@ -2306,6 +2306,13 @@ fail:
 	&& !(w3m_dump & DUMP_EXTRA)) {
 	uncompress_stream(&f, &pu.real_file);
     }
+    /*
+     * TODO(rkta): Something wrong here with DUMP_SOURCE and compressed streams
+     * If DUMP_SOURCE is set and the data is compressed, t will be set to the
+     * type of the compressed data which is not handled in the later if-else
+     * block to determine the function to assign to loadproc. loadproc will be
+     * NULL leading to an error when loading the page.
+    */
     else if (f.compression != CMP_NOCOMPRESS) {
 	if (!(w3m_dump & DUMP_SOURCE) &&
 	    (w3m_dump & ~DUMP_FRAME || is_text_type(t)
@@ -2373,6 +2380,8 @@ fail:
     }
     else if (w3m_dump & DUMP_FRAME)
 	return NULL;
+    else
+	proc = NULL;
 
     if (t_buf == NULL)
 	t_buf = newBuffer(INIT_BUFFER_WIDTH);
